File: go/0459E-重复的子字符串.go
/*
 * @Author: SukiEva
 * @Date: 2021-10-19 20:40:21
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */

package main

func getNext(next []int, s string) {
	i, j := 0, -1
	next[0] = j
	for i < len(s) {
		for j != -1 && s[i] != s[j] { // 前后缀不同，向前回退
			j = next[j]
		}
		i++
		j++
		next[i] = j
	}
}

func repeatedSubstringPattern(s string) bool {
	if len(s) == 0 {
		return false
	}
	next := make([]int, len(s)+1)
	getNext(next, s)
	l := len(s)
	if next[l] != 0 && l%(l-next[l]) == 0 {
		return true
	}
	return false
}

File: cpp/0459E-重复的子字符串.cpp
/*
 * @Author: SukiEva
 * @Date: 2021-12-16 15:41:17
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */
#include <bits/stdc++.h>

#include "include/ListNode.h"
#include "include/TreeNode.h"

using namespace std;

class Solution {
   public:
    bool repeatedSubstringPattern(string s) {
        if (s.size() == 0) return false;
        int len = s.size();
        int nx[len + 1];
        getNext(s, nx);
        if (nx[len] != 0 && len % (len - nx[len]) == 0) return true;
        return false;
    }

    void getNext(const string& s, int* nx) {
        int i = 0;
        int j = nx[0] = -1;
        while (i < s.size()) {
            while (-1 != j && s[i] != s[j]) j = nx[j];
            if (s[++i] == s[++j])
                nx[i] = nx[j];
            else
                nx[i] = j;
        }
    }
};